amaranth_soc/csr/bus.py

```diff
@@ -204,6 +282,59 @@ class Multiplexer(Elaboratable):
     register occupies in the CPU address space, and the width of the CSR bus.
 
+    Buffering
+    ---------
+
+    CSR registers use separate shadows to buffer read and write accesses, which can happen in the
+    same transaction. To save resources, both shadows are shared by every CSR register using them.
+    By default, the contents of a shadow are split in `2 ** ceil(log2(w/n))` chunks of *n*-bits
+    where *n* is the CSR bus data width and *w* the width of the widest register.
+
+    For example, let's consider the read shadow of a multiplexer with a 4-bit wide bus:
+
+    .. code-block::
+
+        +----------+-----------------+
+        |  chunks  | 0 1 2 3 4 5 6 7 |
+        +----------+-----------------+
+        |          | A B C D E F . . | <- a 24-bit register at address 0x0, containing 0xFEDCBA
+        | r_shadow +-----------------+
+        |          | . . . . A B C . | <- a 12-bit register at address 0xC, containing 0xCBA
+        +----------+-----------------+
+
+    While this configuration minimizes the amount of flip-flops in a shadow, having many registers
+    share the same chunks can create a deeply nested mux tree. This may potentially impact timing
+    closure or logic usage, depending on the platform toolchain and technology.
+
+    The amount of chunks of a shadow may also depend on the following:
+        * If a CSR register isn't naturally aligned (i.e. its address isn't a multiple of its
+          size), it will use twice as many shadow chunks.
+        * If more registers share the same chunk than what is allowed by the `shadow_overlaps`
+          parameter, the shadow size is doubled (as many times as needed) in order to balance
+          chunk usage.
+
+    To illustrate the chunk balancing process, let's revisit the above example in the case where
+    `shadow_overlaps` is set to `0` (i.e. a chunk is exclusive to a single register):
+
+    .. code-block::
+
+        +----------+-----------------+-----------------+
+        |  blocks  |        0        |        1        |
+        +----------+-----------------+-----------------+
+        |  chunks  | 0 1 2 3 4 5 6 7 | 8 9 A B C D E F |
+        +----------+-----------------+-----------------+
+        |          | A B C D E F . . | . . . . . . . . | <- a 24-bit register at address 0x0
+        | r_shadow +-----------------+-----------------+
+        |          | . . . . . . . . | . . . . A B C . | <- a 12-bit register at address 0xC
+        +----------+-----------------+-----------------+
+
+    A use case for setting `shadow_overlaps` to `0` would be to loosen the assumption of exclusive
+    ownership of the multiplexer into exclusive ownership of a set of registers (i.e. transactions
+    may be interleaved by multiple initiators, as long as they target different registers).
+
+    Note that unused shadow chunks are not instantiated in the RTL. The addressing scheme described
+    above is the same for the write shadow.
```

Comment:
This is not a criticism of you, but I have to say that this document is borderline impenetrable even to someone who has the context. I think we should still have it, and I think having it written in this particular way is OK, but we really need an uh... TL;DR for it I guess?

It is necessary and useful to have it described but basically no one who is instantiating a CSR mux has it their goal be "learn about chunk balancing".

These are just my broad thoughts about this particular corner of -SoC and not something we need to act right now. I think Allyson may be a lot of help in getting these docs in a friendlier shape some time shortly later.

Reply:
This was originally written as a comment; I made it a docstring to provide context for `shadow_overlaps`. I agree that this just adds cognitive load to -soc users for little gain, since the default value is fine in most cases.

Reply:
Right. To be clear, my view here is that we can just deal with it later. Let's ship the amazingly more efficient multiplexer (which I look forward to using in Glasgow, by the way!) and continue.

Reply:
I also think this is a valid criticism for most of the documentation about address translation primitives in amaranth-soc.

Reply:
Oh, absolutely. The way memory windows work is incredibly esoteric. We will need to have non-reference documentation that describes the "happy path" most designers will take, and its tradeoffs.

Reply:
I have moved the explanation to `Multiplexer._Shadow` docstrings. I think doing so makes it easier to read, and certainly made it easier to write.

This may seem overkill for a private class, but the `shadow_overlaps` parameter is motivated implementation-specific tradeoffs.